core/commands/p2p.go

```diff
@@ -196,149 +266,164 @@ var p2pStreamLsCmd = &cmds.Command{
 			return
 		}
 
-		proto := "/p2p/" + req.Arguments()[0]
-		if n.P2P.CheckProtoExists(proto) {
-			res.SetError(errors.New("protocol handler already registered"), cmdkit.ErrNormal)
-			return
-		}
+		closeAll, _, _ := req.Option("all").Bool()
+		protoOpt, p, _ := req.Option("protocol").String()
+		listenOpt, l, _ := req.Option("listen-address").String()
+		targetOpt, t, _ := req.Option("target-address").String()
 
-		addr, err := ma.NewMultiaddr(req.Arguments()[1])
+		proto := protocol.ID(protoOpt)
+
+		listen, err := ma.NewMultiaddr(listenOpt)
 		if err != nil {
 			res.SetError(err, cmdkit.ErrNormal)
 			return
 		}
 
-		_, err = n.P2P.NewListener(n.Context(), proto, addr)
+		target, err := ma.NewMultiaddr(targetOpt)
 		if err != nil {
 			res.SetError(err, cmdkit.ErrNormal)
 			return
 		}
 
-		// Successful response.
-		res.SetOutput(&P2PListenerInfoOutput{
-			Protocol: proto,
-			Address:  addr.String(),
-		})
-	},
-}
-
-var p2pStreamDialCmd = &cmds.Command{
-	Helptext: cmdkit.HelpText{
-		Tagline: "Dial to a p2p listener.",
-
-		ShortDescription: `
-Establish a new connection to a peer service.
-
-When a connection is made to a peer service the ipfs daemon will setup one
-time TCP listener and return it's bind port, this way a dialing application
-can transparently connect to a p2p service.
-		`,
-	},
-	Arguments: []cmdkit.Argument{
-		cmdkit.StringArg("Peer", true, false, "Remote peer to connect to"),
-		cmdkit.StringArg("Protocol", true, false, "Protocol identifier."),
-		cmdkit.StringArg("BindAddress", false, false, "Address to listen for connection/s (default: /ip4/127.0.0.1/tcp/0)."),
-	},
-	Run: func(req cmds.Request, res cmds.Response) {
-		n, err := getNode(req)
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
+		if !(closeAll || p || l || t) {
+			res.SetError(errors.New("no matching options given"), cmdkit.ErrNormal)
 			return
 		}
 
-		addr, peer, err := ParsePeerParam(req.Arguments()[0])
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
+		if closeAll && (p || l || t) {
+			res.SetError(errors.New("can't combine --all with other matching options"), cmdkit.ErrNormal)
 			return
 		}
 
-		proto := "/p2p/" + req.Arguments()[1]
-
-		bindAddr, _ := ma.NewMultiaddr("/ip4/127.0.0.1/tcp/0")
-		if len(req.Arguments()) == 3 {
-			bindAddr, err = ma.NewMultiaddr(req.Arguments()[2])
-			if err != nil {
-				res.SetError(err, cmdkit.ErrNormal)
-				return
+		match := func(listener p2p.Listener) bool {
+			if closeAll {
+				return true
+			}
+			if p && proto != listener.Protocol() {
+				return false
 			}
+			if l && !listen.Equal(listener.ListenAddress()) {
+				return false
+			}
+			if t && !target.Equal(listener.TargetAddress()) {
+				return false
+			}
+			return true
 		}
 
-		listenerInfo, err := n.P2P.Dial(n.Context(), addr, peer, proto, bindAddr)
-		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+		var closed int
+		for _, listener := range n.P2P.Listeners.Listeners {
```

Comment:
I think we're going to need to hold a lock while iterating over these. We may want to do this in two steps: filter under a lock and then close.